File: DS_A/Python/Dynamic_Programming/knapsack.py
n,x = map(int,input().split()) # n tovarlar soni, x kerakli og'irlik 
l_p = list(map(int,input().split())) # n ta tovar narxi
l_w = list(map(int,input().split())) # n ta tovar og'irligi

baza = [0]*(x+1) #baza shakllantirish

for i in range(0,n):
	pul = l_p[i] #i-tovar narxi
	to = l_w[i] #i-tovar og'irligi 
	
	for j in range(x,pul-1,-1): # takrorlanmaslik uchun pul-1 dan Xqiymatgacha boramiz
		baza[j]=max(baza[j],baza[j-pul]+to) #taklif qilamiz

print(baza[x])

File: DS_A/C++/Dynamic_Programming/knapsack.cpp
#include <iostream>
#include <vector>

using namespace std;

int main(){
	int n,x;
	cin>>n>>x;
	
	vector <int> l_p(n);
	vector <int> l_w(n);
	for(int i = 0; i<n ; i++){
		cin>>l_p[i];
	}
	for(int i = 0; i<n ; i++){
		cin>>l_w[i];
	}
	
	vector <int> baza(x+1,0);
	int pul,to;
	for(int i = 0; i < n; i++){
		pul = l_p[i];
		to = l_w[i];
		for(int j = x; j>=pul; j--){
			baza[j] = max(baza[j],baza[j-pul]+to);
		}
	}
	
	cout<<baza[x];
}
